### PR TITLE
Add logging to sriov-cni

### PR DESCRIPTION
**Pull request by andreaskaris**
Adds both configurable debug logging to stderr and optionally to
file as well as error logging for issue 275.

Reported-at: https://github.com/k8snetworkplumbingwg/sriov-cni/issues/275
Signed-off-by: Andreas Karis <ak.karis@gmail.com>

### CODE REVIEW

**Comment by andreaskaris** [comment]
@mlguerrero12 @SchSeba  Sorry for the random ping - would any of this here make sense? I'm currently working on a customer case and needed to build a debug image to troubleshoot a race condition, and debug logging would have been helpful to me

**Comment by adrianchiris** [comment]
@andreaskaris maybe use : https://github.com/k8snetworkplumbingwg/cni-log

generally im +1 with adding logs to sriov-cni

**Comment by andreaskaris** [comment]
Will do, have a couple of improvements for [k8snetworkplumbingwg/cni-log](https://github.com/k8snetworkplumbingwg/cni-log) though

**Comment by andreaskaris** [comment]
@adrianchiris I submitted a few improvements for https://github.com/k8snetworkplumbingwg/cni-log
It'd be great if you could have a look before I continue work on this one here! Thanks!

**Comment by adrianchiris** [comment]
>@adrianchiris I submitted a few improvements for https://github.com/k8snetworkplumbingwg/cni-log
It'd be great if you could have a look before I continue work on this one here! Thanks!

Done! lets get cni-log changes in first

File: pkg/logging/logging.go
Comment (andreaskaris) [inline pkg/logging/logging.go:22]: needs to be changed to InvalidLevel after the cni-log change merged

**Comment by SchSeba** [comment]
@andreaskaris great to see this PR!

I have a question if we put in debug but we leave the file empty will crio call still work? I need to test this one.
because I try one time to use fmt.print and that mess with the return to crio

**Comment by andreaskaris** [comment]
> @andreaskaris great to see this PR!
> 
> I have a question if we put in debug but we leave the file empty will crio call still work? I need to test this one. because I try one time to use fmt.print and that mess with the return to crio

@SchSeba I have to admit that I don't follow exactly what you mean (my fault, not yours). Can you elaborate a bit more (rough reproducer / test steps) and I'll try it out?

File: pkg/logging/logging.go
Comment (SchSeba) [inline pkg/logging/logging.go:34]: to remove and error message `cni-log: filename is required when logging to stderr is off - will not log anything`
please switch between this 2 lines

meaning first put the setLogFile and then the setLogStderr

**Comment by SchSeba** [comment]
One  more comment can you please add to all the messages the `cni name` or something like that and also `ContainerID, Netns`

so it will be easy to grep stuff if multiple cni calls or looking at the stderr that gets print in multus container

File: cmd/sriov/main.go
Comment (mlguerrero12) [inline cmd/sriov/main.go:238]: shouldn't it be error level?

File: pkg/types/types.go
Comment (mlguerrero12) [inline pkg/types/types.go:53]: let's use logLevel or log_level to be consistent with other parameters

File: cmd/sriov/main.go
Comment (mlguerrero12) [inline cmd/sriov/main.go:48]: I think it's better to create a function to setup the logging. It could be created in config.go. Of course, the debug log should be the only one that remains in cmdAdd and cmdDel.

Comment (mlguerrero12) [reply cmd/sriov/main.go:48]: perhaps, you should only set the logLevel and logFile when these are not ""

File: pkg/logging/logging.go
Comment (mlguerrero12) [inline pkg/logging/logging.go:26]: init of cni-log pkg calls initloger which already sets empty file and enalbe stderr. So, no need to have it here again. If you decide to set error as log level (instead of info as set by cni-log, see comment above). I think this should be the place.

File: pkg/logging/logging.go
Comment (mlguerrero12) [inline pkg/logging/logging.go:41]: no needed if you remove it from init and set in main.go only if it is not empty.

File: pkg/logging/logging.go
Comment (mlguerrero12) [inline pkg/logging/logging.go:47]: so, is it exclusive? only file or only stderr, but not both?

File: pkg/logging/logging.go
Comment (mlguerrero12) [reply pkg/logging/logging.go:41]: if you set the default level in init, then you can do:

if ll:= cnilog.StringToLevel(l), ll != cnilog.InvalidLevel {
cnilog.SetLogLevel(ll)
}

File: docs/configuration-reference.md
Comment (andreaskaris) [reply docs/configuration-reference.md:20]: Because I originally did not want to print any extra information by default with the introduction of the logging mechanism. That's also the reason why I set logging.Info("Cannot load config file from cache"). Let me switch it to Info

File: cmd/sriov/main.go
Comment (andreaskaris) [reply cmd/sriov/main.go:238]: Because I originally did not want to print any extra information by default with the introduction of the logging mechanism.  Let me change this though

File: pkg/logging/logging.go
Comment (andreaskaris) [reply pkg/logging/logging.go:47]: yeah that was my idea to have an "either log to file or log to stderr". The other model would be: "log to stderr and optionally log to file as well" .. or yet another option would be to expose the stderr switch to users as well and have 3 options. But I find the either/or one most intuitive. That said, I'm open to changing it if you strongly favor something else

File: pkg/logging/logging.go
Comment (andreaskaris) [reply pkg/logging/logging.go:26]: I removed init() altogether

File: pkg/logging/logging.go
Comment (mlguerrero12) [reply pkg/logging/logging.go:47]: No, it's fine like it is. Just wanted to check if it was really intended.

File: cmd/sriov/main.go
Comment (andreaskaris) [reply cmd/sriov/main.go:48]: I prefer functions be omnipotent though and not rely too much on the defaults of the underlying library

**Comment by andreaskaris** [comment]
@mlguerrero12 I made a few changes, taking into consideration your comments. I know I didn't address all of your comments but I'd prefer not to rely too much on the underlying cni-log defaults. Lmk what you think, I'm happy to revert/address further comments :-)

File: cmd/sriov/main.go
Comment (mlguerrero12) [reply cmd/sriov/main.go]: fair

File: pkg/config/config.go
Comment (mlguerrero12) [inline pkg/config/config.go:22]: I think that this is a valid error. If we are not able to unmarshall it, the plugin should return an error. What do you think?

Also, no need to have 2 separate functions. You can embed loadLoggingConf here. It's fine if you prefer to it separated.

File: pkg/logging/logging.go
Comment (mlguerrero12) [inline pkg/logging/logging.go:26]: Let's not export the setters anymore. They are only used in this package. Also, let's remove the setters for continaerid, netns and ifname.

File: pkg/config/config.go
Comment (andreaskaris) [reply pkg/config/config.go]: ok to both

File: pkg/logging/logging.go
Comment (andreaskaris) [reply pkg/logging/logging.go]: done

**Comment by mlguerrero12** [comment]
thank you @andreaskaris.

LGTM

**Review by mlguerrero12 — APPROVED** [review]

**Comment by andreaskaris** [comment]
@zeeke can you have a look? Thanks!

File: pkg/logging/logging.go
Comment (zeeke) [inline pkg/logging/logging.go:56]: nit: This function is not used in this project, and I don't see any specific value in exporting it as public. Can we get rid of it?

Same for  `SetNetNS` and `SetIFName`?

**Review by zeeke — APPROVED** [review]
Left a minor comment, not blocking the PR for that.

LGTM

File: Makefile
Comment (Eoghan1232) [inline Makefile:17]: IMAGE_BUILDER ?= docker

I would prefer something like this, check if the env var is set, if so, take the value, otherwise default.
I would prefer docker to be the default as well.

WDYT?

Comment (andreaskaris) [reply Makefile]: done: https://github.com/k8snetworkplumbingwg/sriov-cni/pull/276/commits/fdbccf1feb606d58661debca58dab505550dc143

**Review by Eoghan1232 — APPROVED** [review]
LGTM!

**Comment by Eoghan1232** [comment]
@zeeke are we good to merge?

**Pull request merged**
